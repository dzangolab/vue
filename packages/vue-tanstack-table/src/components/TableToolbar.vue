<template>
  <div class="toolbar">
    <ButtonElement
      v-if="showResetButton"
      :label="resetButtonLabel"
      variant="outlined"
      severity="secondary"
      @click="$emit('on:reset')"
    />

    <Popup v-if="showColumnAction">
      <ButtonElement
        :label="columnActionButtonLabel"
        variant="outlined"
        severity="secondary"
      />

      <template #content>
        <SortableList :list="items" @on-drag="onDrag" />
      </template>
    </Popup>

    <slot />
  </div>
</template>

<script lang="ts">
export default {
  name: "TableToolbar",
};
</script>

<script setup lang="ts">
import { Checkbox } from "@dzangolab/vue3-form";
import { ButtonElement, Popup } from "@dzangolab/vue3-ui";
import { computed, h } from "vue";

import type { Table } from "@tanstack/vue-table";
import type { VNode } from "vue";

type List = {
  id: number | string;
  data: string;
  render?: (data: unknown) => VNode;
};

const props = defineProps({
  columnActionButtonLabel: {
    default: "Columns",
    type: String,
  },
  hasActionsRow: Boolean,
  hasSelectionRow: Boolean,
  resetButtonLabel: {
    default: "Reset all",
    type: String,
  },
  showColumnAction: Boolean,
  showResetButton: Boolean,
  table: {
    required: true,
    type: Object as () => Table<unknown>,
  },
});

defineEmits(["on:reset"]);

const items = computed(() =>
  props.table
    .getAllLeafColumns()
    .filter((column) => column.id !== "select" && column.id !== "actions")
    .map((column, index) => ({
      id: index,
      data: column,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      render: (data: any) => {
        let header = data.columnDef.header;

        if (typeof data.columnDef.header === "function") {
          header = data.columnDef.header();
        }

        return h(Checkbox, {
          label: header,
          modelValue: data.getIsVisible(),
          "onUpdate:modelValue": () => data.toggleVisibility(),
        });
      },
    })),
);

const onDrag = (sorted: List[]) => {
  props.table.setColumnOrder([
    ...(props.hasSelectionRow ? ["select"] : []),
    ...sorted.map((item: List) => String(item?.id)),
    ...(props.hasActionsRow ? ["actions"] : []),
  ]);
};
</script>
