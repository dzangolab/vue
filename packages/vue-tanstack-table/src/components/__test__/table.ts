import {
  getCoreRowModel,
  getFilteredRowModel,
  getPaginationRowModel,
  getSortedRowModel,
  useVueTable,
} from "@tanstack/vue-table";

export const mockedTable = useVueTable({
  columns: [
    {
      accessorKey: "name",
      enableColumnFilter: true,
      header: "Name",
    },
    {
      accessorKey: "age",
      header: "Age",
      tooltip: true,
    },
    {
      accessorKey: "country",
      header: "Country",
    },
  ],
  data: [
    {
      age: 30,
      country: "USA",
      name: "John Doe",
    },
    {
      age: 25,
      country: "Canada",
      name: "Jane Smith",
    },
    {
      age: 28,
      country: "UK",
      name: "Alice Johnson",
    },
],
  getCoreRowModel: getCoreRowModel(),
  getFilteredRowModel: getFilteredRowModel(),
  getPaginationRowModel: getPaginationRowModel(),
  getSortedRowModel: getSortedRowModel(),
});