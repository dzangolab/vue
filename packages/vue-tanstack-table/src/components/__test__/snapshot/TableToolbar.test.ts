import { mount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";

import { mockedTable } from "../table";
import TableToolbar from "../../TableToolbar.vue";
import { SortableList } from "@dzangolab/vue3-ui";

describe("TableToolbar", () => {
  it("matches snapshot", () => {
    const wrapper = mount(TableToolbar, {
      global: {
        stubs: {
          SortableList: true,
        }
      },
      props: {
        showColumnAction: true,
        showResetButton: true,
        table: mockedTable,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
