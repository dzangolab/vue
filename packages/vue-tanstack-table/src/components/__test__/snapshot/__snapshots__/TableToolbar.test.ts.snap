// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableToolbar > matches snapshot 1`] = `
<div
  class="toolbar"
>
  <button
    aria-label="button"
    class="button secondary medium outlined"
    role="button"
  >
    <!---->
    <div
      class="label"
    >
      
      <span>
        Reset all
      </span>
      
    </div>
    <!---->
    <!---->
  </button>
  <div
    class="popup-container"
  >
    <div
      aria-expanded="false"
      aria-label="popup"
      class="popup-trigger"
    >
      
      <button
        aria-label="button"
        class="button secondary medium outlined"
        role="button"
      >
        <!---->
        <div
          class="label"
        >
          
          <span>
            Columns
          </span>
          
        </div>
        <!---->
        <!---->
      </button>
      
    </div>
    <!---->
  </div>
  
  
</div>
`;
